File: bloock-bridge/src/entity_mappings/proof.rs
use bloock_core::{
    from_hex,
    proof::entity::{anchor::ProofAnchor as ProofAnchorCore, proof::Proof as ProofCore},
    to_hex, H256,
};

use crate::{
    error::{BridgeError, BridgeResult},
    items::{Proof, ProofAnchor},
};

impl From<ProofCore> for Proof {
    fn from(p: ProofCore) -> Self {
        Self {
            leaves: p.leaves.iter().map(|&l| to_hex(l)).collect(),
            nodes: p.nodes.iter().map(|&n| to_hex(n)).collect(),
            depth: p.depth,
            bitmap: p.bitmap,
            anchor: Some(p.anchor.into()),
        }
    }
}

impl TryFrom<Proof> for ProofCore {
    type Error = BridgeError;
    fn try_from(p: Proof) -> BridgeResult<Self> {
        let anchor = match p.anchor {
            Some(anchor) => anchor,
            None => return Err(BridgeError::MissingAnchor),
        };
        Ok(ProofCore {
            leaves: p
                .leaves
                .iter()
                .map(|l| {
                    from_hex(l).map_err(|_| {
                        BridgeError::RequestDeserialization(
                            "couldn't deserialize leaves".to_string(),
                        )
                    })
                })
                .collect::<BridgeResult<Vec<H256>>>()?,
            nodes: p
                .nodes
                .iter()
                .map(|l| {
                    from_hex(l).map_err(|_| {
                        BridgeError::RequestDeserialization(
                            "couldn't deserialize nodes".to_string(),
                        )
                    })
                })
                .collect::<BridgeResult<Vec<H256>>>()?,
            depth: p.depth,
            bitmap: p.bitmap,
            anchor: anchor.into(),
        })
    }
}

impl From<ProofAnchorCore> for ProofAnchor {
    fn from(p: ProofAnchorCore) -> Self {
        Self {
            anchor_id: p.anchor_id,
            networks: p
                .networks
                .iter()
                .map(|network| network.clone().into())
                .collect(),
            root: p.root,
            status: p.status,
        }
    }
}

impl From<ProofAnchor> for ProofAnchorCore {
    fn from(p: ProofAnchor) -> Self {
        Self {
            anchor_id: p.anchor_id,
            networks: p
                .networks
                .iter()
                .map(|network| network.clone().into())
                .collect(),
            root: p.root,
            status: p.status,
        }
    }
}

File: languages/go/client/entity/proof.go
package entity

import "github.com/bloock/go-bridge/internal/bridge/proto"

type Proof struct {
	leaves []string
	nodes  []string
	depth  string
	bitmap string
	anchor ProofAnchor
}

func NewProofFromProto(p *proto.Proof) Proof {
	return Proof{
		leaves: p.Leaves,
		nodes:  p.Nodes,
		depth:  p.Depth,
		bitmap: p.Bitmap,
		anchor: NewProofAnchorFromProto(p.Anchor),
	}
}

func (p Proof) ToProto() *proto.Proof {
	return &proto.Proof{
		Leaves: p.leaves,
		Nodes:  p.nodes,
		Depth:  p.depth,
		Bitmap: p.bitmap,
		Anchor: p.anchor.ToProto(),
	}
}

type ProofAnchor struct {
	AnchorID int64
	Networks []AnchorNetwork
	Root     string
	Status   string
}

func NewProofAnchorFromProto(p *proto.ProofAnchor) ProofAnchor {
	return ProofAnchor{
		AnchorID: p.AnchorId,
		Networks: MapAnchorNetworksFromProto(p.Networks),
		Root:     p.Root,
		Status:   p.Status,
	}
}

func (p ProofAnchor) ToProto() *proto.ProofAnchor {
	networks := make([]*proto.AnchorNetwork, len(p.Networks))
	for i, network := range p.Networks {
		networks[i] = network.ToProto()
	}

	return &proto.ProofAnchor{
		AnchorId: p.AnchorID,
		Networks: networks,
		Root:     p.Root,
		Status:   p.Status,
	}
}
